save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/service/GitPreprocessorService.kt

```diff
@@ -0,0 +1,126 @@
+package com.saveourtool.save.preprocessor.service
+
+import com.saveourtool.save.entities.GitDto
+import com.saveourtool.save.preprocessor.config.ConfigProperties
+import com.saveourtool.save.preprocessor.utils.cloneToDirectory
+import com.saveourtool.save.utils.*
+import org.eclipse.jgit.util.FileUtils
+import org.slf4j.Logger
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Mono
+import reactor.core.scheduler.Schedulers
+import java.io.IOException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.time.Instant
+import kotlin.io.path.absolutePathString
+import kotlin.io.path.createDirectories
+
+typealias GitRepositoryProcessor<T> = (Path, Instant) -> Mono<T>
+typealias ArchiveProcessor<T> = (Path) -> Mono<T>
+
+/**
+ * Additional service for Git based [com.saveourtool.save.entities.TestSuitesSource]s
+ */
+@Service
+class GitPreprocessorService(
+    configProperties: ConfigProperties,
+) {
+    private val workingDir = Paths.get(configProperties.repository, "tmp")
+        .also { it.createDirectories() }
+
+    private fun createTempDirectoryForRepository() = Files.createTempDirectory(
+        workingDir,
+        "repository-"
+    )
+
+    private fun createTempTarFile() = Files.createTempFile(
+        workingDir,
+        "archive-",
+        ARCHIVE_EXTENSION
+    )
+
+    /**
+     * @param gitDto
+     * @param branch
+     * @param sha1
+     * @param repositoryProcessor operation on folder should be finished here -- folder will be removed after it
+     * @return result of [repositoryProcessor]
+     * @throws IllegalStateException
+     * @throws Exception
+     */
+    @Suppress("TooGenericExceptionCaught")
+    fun <T> cloneAndProcessDirectory(
+        gitDto: GitDto,
+        branch: String,
+        sha1: String,
+        repositoryProcessor: GitRepositoryProcessor<T>,
+    ): Mono<T> {
+        val cloneAction: () -> Pair<Path, Instant> = {
+            val tmpDir = createTempDirectoryForRepository()
```

Comment:
is it possible to move this code inside of Mono?

Reply:
it's a callable -- initialization of Mono.  

I use here `Mono.usingWhen` to control a tmp file which should be removed after all operations on this tmp file (kdoc describes it)